### PR TITLE
[RELAY][PASS] FoldScaleAxis Backward

### PR DESCRIPTION
**Pull request by tqchen**
This is a followup of #2020 , this PR implemented the infrastructure to do backward folding of scale on axis.

## Goal
Fold the scaling of axis(usually caused by BatchNorm) into weight of conv2d in the past. For example

Old:
```
%1 = conv2d(%1, %w, data_layout="NHWC")
%2 = multiply(%x, %scale)
```
Transformed:
```
# scale weight's output channel
%1 = multiply(%w, expand_dims(%scale, axis=1, num_newaxis=3))
%2 = conv2d(%x, %1, data_layout="NHWC")
```

Further constant folding can fold the multiplication and we remove the scaling in the network.

## The Algorithm
The general idea is similar to the Forward algorithm.  We pass additional argument ```(axes, scale)```, when doing transformation, to request the result to satisfy
```
result = value
for i, k in enumerate(axes):
   k-ith dimension of result *= i-th dimension of scale
```
The problem is again we don;t want to blindly propagate scale backward it won't get consumed. So we run a forward "preparation phase", which propagates the demand of the potential axes scaling.

The new pass is more general than the FoldScaleAxis in nnvm
- The new pass support arbitrary scaling of multiple axes(although the further implementation is necessary), which could be helpful in NHWCc case,
- The new pass support folding backward into sum of two of conv2d, which was not possible in NNVM.

### CODE REVIEW

File: src/relay/pass/pattern_util.h [hunk 1]
```diff
@@ -100,11 +101,31 @@ inline Expr ExpandBiasToMatchAxis(Expr bias,
...
+ */
+inline bool IsDepthwiseConv2D(const Call& call,
+                              const Conv2DAttrs* param,
+                              const Layout& weight_layout) {
+  static const Layout kOIHW("OIHW");
+  auto wshape = ConvertLayout(
+      call->args[1]->type_as<TensorTypeNode>()->shape,
+      weight_layout, kOIHW);
+  return is_const_int(wshape[0], param->groups) &&
+      is_const_int(wshape[1], 1);
```
Comment (FrozenGene) [inline src/relay/pass/pattern_util.h]: Could we use ```is_const_int(wshape[0], param->groups)``` to check depthwise convolution? In theory, wshape[1] could be not equal to 1, although TF / MXNet / Caffe and so on frontends will make it be 1 even though the depth channel multiplier not  be 1 (such as 0.25)

File: src/relay/pass/fold_scale_axis.cc [hunk 2]
```diff
@@ -542,13 +550,416 @@ RELAY_REGISTER_OP("nn.conv2d")
...
+                           const BackwardTransformer& transformer) {
+  if (!axes.defined()) {
+    return transformer->NormalCallTransform(call.operator->());
+  }
+  Expr input = transformer->Transform(
+      call->args[0], axes, scale);
+  return CallNode::make(call->op, {input}, call->attrs, call->type_args);
+}
+
+RELAY_REGISTER_OP("nn.relu")
```
Comment (FrozenGene) [inline src/relay/pass/fold_scale_axis.cc]: How about add `clip` operator like `relu`? Like Tensorflow frontend or Keras frontend, which use `clip` to implement `relu6` operator.

File: src/relay/pass/pattern_util.h (hunk 1 above)
Comment (tqchen) [reply src/relay/pass/pattern_util.h]: The current code makes the assumption of input and output groups being even, and the special logic relies on this

 A better approach would enhance the Conv2D code itself to handle general group convolution, which should not be too hard as a followup contribution from anyone in the community:)

File: src/relay/pass/fold_scale_axis.cc (hunk 2 above)
Comment (tqchen) [reply src/relay/pass/fold_scale_axis.cc]: clip can be supported, this PR mainly introduces the scaffolding of things necessary, with  new folder infrastructure, we can support clip as well as other operators(flatten, transpose) by more registrations

File: src/relay/pass/fold_scale_axis.cc [hunk 3]
```diff
@@ -41,9 +41,14 @@ using runtime::TypedPackedFunc;
 // we run a backward "preparation phase", which propagates the demand
 // of the potential axes scaling back to its input.
 //
-// The folding process is done in two steps:
+// Forward folding process is done in two steps:
 // - Prepare phase: backward propagation of demand.
 // - Transform phase: forward transformation,
+//
+// Similarly, borward folding process is done in two steps:
```
Comment (masahi) [inline src/relay/pass/fold_scale_axis.cc]: backward

File: src/relay/pass/fold_scale_axis.cc [hunk 4]
```diff
@@ -542,13 +550,416 @@ RELAY_REGISTER_OP("nn.conv2d")
...
 TVM_REGISTER_API("relay._ir_pass.forward_fold_scale_axis")
 .set_body_typed<Expr(Expr)>(ForwardFoldScaleAxis);
 
+//----------------------------------------
+// Implement backward transformations.
+//----------------------------------------
+class BackwardTransformer;
+
+/*!
+ * \brief Preparation function for for pass scale backward.
```
Comment (masahi) [inline src/relay/pass/fold_scale_axis.cc]: for

**Comment by tqchen** [comment]
Thanks, @masahi @FrozenGene  I have updated the PR according to your comments

**Comment by tqchen** [comment]
Thanks @FrozenGene @masahi , this is merged. Followup PRs to support more ops(clip) and grouped conv is more than welcomed!
